model/device.go

```diff
@@ -69,6 +69,64 @@ type DeviceAttribute struct {
 	Timestamp   *time.Time  `json:"timestamp,omitempty" bson:",omitempty"`
 }
 
+func (da DeviceAttributes) GetByName(name string) *DeviceAttribute {
+	for _, attribute := range da {
+		if attribute.Name == name {
+			rc := attribute
+			return &rc
+		}
+	}
+	return nil
+}
+
+func (da DeviceAttributes) ToMap(excludeNames ...string) map[string]DeviceAttribute {
+	rc := make(map[string]DeviceAttribute, len(da))
+	for _, v := range da {
+		exclude := false
+		for _, excluded := range excludeNames {
+			if v.Name == excluded {
+				exclude = true
+				break
+			}
+		}
+		if exclude {
+			continue
+		}
+		rc[v.Scope+"_"+v.Name] = v
+	}
+	return rc
+}
+
+func (da DeviceAttributes) Equal(existingAttributesArray DeviceAttributes) bool {
+	// cannot be equal if they are of different size (the cheapest check first)
+	if len(da)-2 != len(existingAttributesArray) { // -2 comes from updated and created timestamps
+		return false
+	}
+
+	// cannot be equal if any of the values differ (the most common case)
+	attributes := da.ToMap(AttrNameUpdated, AttrNameCreated)
+	existingAttributes := existingAttributesArray.ToMap(AttrNameUpdated, AttrNameCreated)
+	if len(attributes) != len(existingAttributes) {
+		return false
+	}
+	for k, v := range attributes {
+		if e, ok := existingAttributes[k]; ok {
+			if !v.Equal(e) {
+				return false
+			}
+		}
+	}
+
+	// cannot be equal if keys differ (the last possibility)
+	for k := range attributes {
+		if _, ok := existingAttributes[k]; !ok {
+			return false
+		}
+	}
```

Comment:
You can collapse the second loop in the first one.
```suggestion
	for k, v := range attributes {
		if e, ok := existingAttributes[k]; ok {
			if !v.Equal(e) {
				return false
			}
		} else {
			return false
		}
	}
```

Reply:
this was on purpose, since it is the less common scenario. but ok, no problem.